WalletWasabi/Blockchain/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -311,15 +311,16 @@ private ProcessedResult ProcessNoLock(SmartTransaction tx)
 		&& !weAreAmongTheSender // we are not one of the senders (it is not a self-spending tx or coinjoin)
 		&& Coins.Any(c => c.HdPubKey == hdPubKey); // the destination address has already been used (address reuse)
 
-	private static void SaveInternalKeysLatestSpendingHeight(Height txHeight, IEnumerable<HdPubKey> internalKeys)
+	private static void SaveInternalKeysLatestSpendingHeight(Height txHeight, ImmutableList<SmartCoin> myInputs)
 	{
-		foreach (var spenderKey in internalKeys)
+		IEnumerable<IGrouping<HdPubKey, SmartCoin>> grouping = myInputs
+			.Where(c => c.HdPubKey.IsInternal) // Only internal ones.
+			.Where(c => !c.IsSpent()) // Remove all spent coins.
```

Comment:
Maybe it is just me but I found this comment confusing compared to the one above it.
To be consistent how about this:

```suggestion
			.Where(c => !c.IsSpent()) // Only unspent coins.
```